File: android/src/main/kotlin/io/sentient/android/settings/voice/VoiceViewModel.kt
// ---------------------------------------------------------------------------
// VoiceViewModel — thin state-holder for the Voice settings list. Mirrors the webui
// VoicesPanel: load the shared library once, filter client-side (search / source /
// tags / language), preview / pick / delete per pack, and gate the "Clone from Fish"
// entry on the fish-browse feature flag.
//
// The active voice id comes from a direct profile read — SettingsComponent exposes
// its repos "for the rare direct read", and there is no dedicated active-voice
// usecase (webui reads profile.voice.id the same way). Every op folds the result
// envelope / FishResult exhaustively; audio bytes are never logged (usecase-side).
// ---------------------------------------------------------------------------
package io.sentient.android.settings.voice

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.data.settings.ProfileRepository
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ObserveSettingsAccessUseCase
import io.sentient.mobiledata.usecase.settings.VoiceFilterState
import io.sentient.mobiledata.usecase.settings.VoicesUseCases
import io.sentient.mobiledata.usecase.settings.deriveLanguageOptions
import io.sentient.mobiledata.usecase.settings.deriveTagOptions
import io.sentient.mobiledata.usecase.settings.filterVoices
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.VoiceSummary
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Render state for the Voice list page. `shown` is the filtered view of `allVoices`. */
data class VoiceUiState(
    val loading: Boolean = true,
    val errorMessage: String? = null,
    val allVoices: List<VoiceSummary> = emptyList(),
    val shown: List<VoiceSummary> = emptyList(),
    val allTags: List<String> = emptyList(),
    val allLanguages: List<String> = emptyList(),
    val activeVoiceId: String = "",
    val filter: VoiceFilterState = VoiceFilterState(),
    val fishEnabled: Boolean = false,
    val previewLoadingId: String? = null,
    val previewPlayingId: String? = null,
    val busyId: String? = null,
    val pendingDelete: VoiceSummary? = null,
    val notice: String? = null,
)

class VoiceViewModel(
    private val voices: VoicesUseCases,
    private val profile: ProfileRepository,
    private val observeAccess: ObserveSettingsAccessUseCase,
    private val player: VoicePreviewPlayer,
) : ViewModel() {
    private val log = createLogger("android", "settings", "voice-vm")

    private val _state = MutableStateFlow(VoiceUiState())
    val state: StateFlow<VoiceUiState> = _state.asStateFlow()

    init {
        refresh()
        loadAccess()
    }

    /** Reload the library + active voice; recompute the filtered view + facet options. */
    fun refresh() {
        _state.update { it.copy(loading = true, errorMessage = null) }
        viewModelScope.launch {
            val active = loadActiveVoiceId()
            when (val r = voices.listFiltered(VoiceFilterState())) {
                is SentientResult.Success -> applyLoaded(r.data, active)
                is SentientResult.Failure -> {
                    log.warn("list.failed", mapOf("kind" to r.error.kind))
                    _state.update { it.copy(loading = false, errorMessage = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    private fun applyLoaded(all: List<VoiceSummary>, active: String?) {
        log.info("list.loaded", mapOf("count" to all.size))
        _state.update { s ->
            s.copy(
                loading = false,
                allVoices = all,
                shown = filterVoices(all, s.filter),
                allTags = deriveTagOptions(all),
                allLanguages = deriveLanguageOptions(all),
                activeVoiceId = active ?: s.activeVoiceId,
            )
        }
    }

    private suspend fun loadActiveVoiceId(): String? =
        when (val p = profile.getProfile()) {
            is SentientResult.Success -> p.data.voice.id
            else -> null
        }

    private fun loadAccess() {
        viewModelScope.launch {
            when (val r = observeAccess()) {
                is SentientResult.Success -> _state.update { it.copy(fishEnabled = r.data.fishBrowseEnabled) }
                else -> Unit
            }
        }
    }

    fun setQuery(q: String) = updateFilter { it.copy(query = q) }

    fun setSource(source: String) = updateFilter { it.copy(source = source) }

    fun setLanguage(language: String) = updateFilter { it.copy(language = language) }

    fun toggleTag(tag: String) =
        updateFilter { f -> f.copy(tags = if (tag in f.tags) f.tags - tag else f.tags + tag) }

    private fun updateFilter(transform: (VoiceFilterState) -> VoiceFilterState) {
        _state.update { s ->
            val next = transform(s.filter)
            s.copy(filter = next, shown = filterVoices(s.allVoices, next))
        }
    }

    /** Play (or stop) a live-synth preview for [voice] in the pack's own language. */
    fun togglePreview(voice: VoiceSummary) {
        val s = _state.value
        if (s.previewPlayingId == voice.voiceId || s.previewLoadingId == voice.voiceId) {
            player.stop()
            _state.update { it.copy(previewPlayingId = null, previewLoadingId = null) }
            return
        }
        _state.update { it.copy(previewLoadingId = voice.voiceId) }
        viewModelScope.launch {
            when (val r = voices.preview(voice.voiceId, voice.language)) {
                is SentientResult.Success -> {
                    _state.update { it.copy(previewLoadingId = null, previewPlayingId = voice.voiceId) }
                    player.playBytes(r.data) { onPreviewEnded(voice.voiceId) }
                }
                is SentientResult.Failure -> {
                    log.warn("preview.failed", mapOf("kind" to r.error.kind))
                    _state.update { it.copy(previewLoadingId = null, notice = "Couldn't play preview") }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    private fun onPreviewEnded(id: String) {
        _state.update { if (it.previewPlayingId == id) it.copy(previewPlayingId = null) else it }
    }

    /** Set [voice] as the active reply voice (fast op — no restart). */
    fun pick(voice: VoiceSummary) {
        if (voice.voiceId == _state.value.activeVoiceId) return
        _state.update { it.copy(busyId = voice.voiceId) }
        viewModelScope.launch {
            when (val r = voices.pickActive(voice.voiceId)) {
                is SentientResult.Success ->
                    _state.update { it.copy(busyId = null, activeVoiceId = r.data.voice.id) }
                is SentientResult.Failure -> {
                    log.warn("pick.failed", mapOf("kind" to r.error.kind))
                    _state.update { it.copy(busyId = null, notice = "Couldn't switch voice") }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun requestDelete(voice: VoiceSummary) = _state.update { it.copy(pendingDelete = voice) }

    fun cancelDelete() = _state.update { it.copy(pendingDelete = null) }

    /** Delete the pack awaiting confirmation, then refetch the list. */
    fun confirmDelete() {
        val target = _state.value.pendingDelete ?: return
        _state.update { it.copy(pendingDelete = null, busyId = target.voiceId) }
        viewModelScope.launch {
            when (val r = voices.delete(target.voiceId)) {
                is SentientResult.Success -> {
                    _state.update { it.copy(busyId = null, notice = "Voice deleted") }
                    refresh()
                }
                is SentientResult.Failure -> {
                    log.warn("delete.failed", mapOf("kind" to r.error.kind))
                    _state.update { it.copy(busyId = null, notice = "Couldn't delete voice") }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun clearNotice() = _state.update { it.copy(notice = null) }

    override fun onCleared() {
        player.release()
    }
}

File: ios/App/Settings/Voice/VoiceViewModel.swift
// ---------------------------------------------------------------------------
// VoiceViewModel — the thin state-holder for the Voice list page. Loads the voice
// library + the current active voice, owns the filter inputs (search / source /
// tags / language), and drives preview / pick / delete over `settings.voices`.
//
// The library is fetched ONCE (unfiltered) and filtered locally for responsive
// search — mirroring the webui VoicesPanel (fetch-all-once + `filterPacks` in the
// view), and keeping the tag/language filter options derivable from the full set.
// The KMP `filterVoices` stays the wire-tested source of truth; this Swift filter
// is a view-layer mirror for keystroke responsiveness only.
//
// Active-voice seed reads `profileRepository.getProfile()` directly — the
// SettingsComponent exposes its repos "for the rare direct read" (its own doc);
// VoicesUseCases has no plain active-voice getter, and `pickActive` returns the
// saved profile so the badge re-anchors from the write.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

extension Notification.Name {
    /// Posted by the Add / Fish-clone VMs after a successful create/clone so the
    /// still-alive VoiceScreen refetches its list when the sub-page pops back.
    /// (The parent's NavigationStack `.task` never re-fires on pop, and the host
    /// route wiring can't be extended with a success callback — this is the
    /// deterministic cross-screen signal.)
    static let voiceLibraryChanged = Notification.Name("io.sentient.voiceLibraryChanged")
}

@MainActor
@Observable
final class VoiceViewModel {
    enum Phase: Equatable { case loading, loaded, failed }

    private(set) var phase: Phase = .loading
    private(set) var allVoices: [VoiceSummary] = []
    private(set) var activeVoiceId: String = ""
    /// Gates the "Clone from Fish" entry (features.fish_browse_enabled). Degrades
    /// to hidden on any access-read failure (fail-safe) — mirrors the root gate.
    private(set) var fishBrowseEnabled = false

    // Filter inputs (mirror VoiceFilterState; "all"/"" = no filter).
    var query: String = ""
    var source: String = "all"
    var selectedTags: [String] = []
    var language: String = ""

    // Ephemeral op state.
    private(set) var previewingId: String?
    private(set) var previewLoadingId: String?
    private(set) var busy = false
    var notice: String?
    var pendingDelete: VoiceSummary?

    private let settings: SettingsComponent
    private let player = VoiceSamplePlayer()
    private var previewTask: Task<Void, Never>?
    private let log = AppLog("settings", "voice-vm")

    init(settings: SettingsComponent) {
        self.settings = settings
        player.onFinished = { [weak self] in self?.previewingId = nil }
    }

    var shownVoices: [VoiceSummary] {
        VoiceLibraryFiltering.apply(
            allVoices, query: query, source: source,
            selectedTags: selectedTags, language: language
        )
    }

    var tagOptions: [String] {
        Array(Set(allVoices.flatMap { $0.tags })).sorted()
    }

    var languageOptions: [(code: String, label: String)] {
        VoiceLanguages.filterOptions(present: allVoices.map { $0.language })
    }

    /// Load the library + active voice + fish gate. Idempotent; safe on each `.task`.
    func load() async {
        log.info("load")
        phase = allVoices.isEmpty ? .loading : phase
        async let profile: Void = seedActiveVoice()
        async let access: Void = loadFishGate()
        await refresh()
        await profile
        await access
    }

    private func loadFishGate() async {
        do {
            let result = try await settings.observeSettingsAccess.invoke()
            if case .success(let s) = onEnum(of: result) {
                fishBrowseEnabled = s.data.fishBrowseEnabled
                log.info("fishGate enabled=\(fishBrowseEnabled)")
            }
        } catch is CancellationError {
        } catch {
            log.warn("fishGate.threw code=transport")
        }
    }

    /// Re-fetch the list + active voice after a sub-page create/clone (clone
    /// auto-activates server-side, so the badge must re-anchor too).
    func onLibraryChanged() async {
        log.info("library.changed")
        await refresh()
        await seedActiveVoice()
    }

    /// Re-fetch just the library list (after create/delete/clone returns).
    func refresh() async {
        do {
            let result = try await settings.voices.listFiltered(filter: emptyFilter())
            switch onEnum(of: result) {
            case .success(let s):
                allVoices = (s.data as? [VoiceSummary]) ?? []
                phase = .loaded
                log.info("load.ok count=\(allVoices.count)")
            case .failure(let f):
                phase = .failed
                log.warn("load.failed kind=\(f.error.kind.name)")
            case .loading:
                phase = .loading
            }
        } catch is CancellationError {
        } catch {
            phase = .failed
            log.warn("load.threw code=transport")
        }
    }

    private func seedActiveVoice() async {
        do {
            let result = try await settings.profileRepository.getProfile()
            if case .success(let s) = onEnum(of: result) {
                activeVoiceId = s.data.voice.id
                log.debug("active.seed")
            }
        } catch is CancellationError {
        } catch {
            log.warn("active.seed.threw code=transport")
        }
    }

    /// Toggle preview for [voice]: stop if it's playing, else fetch + play its WAV.
    func togglePreview(_ voice: VoiceSummary) {
        if previewingId == voice.voiceId {
            player.stop()
            previewingId = nil
            return
        }
        previewLoadingId = voice.voiceId
        previewTask?.cancel()
        previewTask = Task { await playPreview(voice) }
    }

    private func playPreview(_ voice: VoiceSummary) async {
        defer { if previewLoadingId == voice.voiceId { previewLoadingId = nil } }
        do {
            let result = try await settings.voices.preview(voiceId: voice.voiceId, lang: voice.language)
            guard previewLoadingId == voice.voiceId else { return } // superseded
            switch onEnum(of: result) {
            case .success(let s):
                let data = s.data.toData()
                log.debug("preview.ok bytes=\(data.count)")
                try player.playData(data)
                previewingId = voice.voiceId
            case .failure(let f):
                notice = "Couldn't play preview"
                log.warn("preview.failed kind=\(f.error.kind.name)")
            case .loading:
                break
            }
        } catch {
            notice = "Couldn't play preview"
            log.warn("preview.threw code=transport")
        }
    }

    /// Pick [voice] as the active voice (FAST profile PUT; re-anchors the badge).
    func pick(_ voice: VoiceSummary) {
        guard !busy, activeVoiceId != voice.voiceId else { return }
        busy = true
        Task {
            defer { busy = false }
            do {
                let result = try await settings.voices.pickActive(voiceId: voice.voiceId)
                switch onEnum(of: result) {
                case .success(let s):
                    activeVoiceId = s.data.voice.id
                    log.info("pick.ok")
                case .failure(let f):
                    notice = "Couldn't switch voice"
                    log.warn("pick.failed kind=\(f.error.kind.name)")
                case .loading:
                    break
                }
            } catch {
                notice = "Couldn't switch voice"
                log.warn("pick.threw code=transport")
            }
        }
    }

    /// Confirm-then-delete a user pack. The server resets the active pick to the
    /// default when the deleted pack was active; we re-seed from the fresh profile.
    func confirmDelete() {
        guard let voice = pendingDelete else { return }
        pendingDelete = nil
        busy = true
        if previewingId == voice.voiceId { player.stop(); previewingId = nil }
        Task {
            defer { busy = false }
            do {
                let result = try await settings.voices.delete(voiceId: voice.voiceId)
                switch onEnum(of: result) {
                case .success(let s):
                    notice = s.data.warning != nil ? "Voice deleted, but the active pick may be stale." : "Voice deleted"
                    log.info("delete.ok warning=\(s.data.warning != nil)")
                    await refresh()
                    await seedActiveVoice()
                case .failure(let f):
                    notice = "Couldn't delete voice"
                    log.warn("delete.failed kind=\(f.error.kind.name)")
                case .loading:
                    break
                }
            } catch {
                notice = "Couldn't delete voice"
                log.warn("delete.threw code=transport")
            }
        }
    }

    /// Stop playback + release the audio session on screen teardown.
    func teardown() {
        previewTask?.cancel()
        previewTask = nil
        player.stop()
        previewingId = nil
        previewLoadingId = nil
    }

    func toggleTag(_ tag: String) {
        if let idx = selectedTags.firstIndex(of: tag) {
            selectedTags.remove(at: idx)
        } else {
            selectedTags.append(tag)
        }
    }

    private func emptyFilter() -> VoiceFilterState {
        VoiceFilterState(query: "", source: "all", tags: [], language: "")
    }
}
